File: training-data/matlab/collatz_sequence.m
function seq = collatz_sequence(n)
    seq = n;
    while n ~= 1
        if mod(n, 2) == 0
            n = n / 2;
        else
            n = 3 * n + 1;
        end
        seq(end + 1) = n;
    end
end

disp(collatz_sequence(6))
disp(numel(collatz_sequence(27)))

File: training-data/prolog/collatz_sequence.pl
% The Collatz (hailstone) sequence: halve an even number, or triple it
% and add one if odd, repeating until reaching 1. Conjectured to always
% terminate, though no proof is known.
collatz(1, [1]) :- !.
collatz(N, [N|Rest]) :-
    N > 1,
    ( 0 is N mod 2 -> N1 is N // 2 ; N1 is 3 * N + 1 ),
    collatz(N1, Rest).

collatz_length(N, Length) :- collatz(N, Seq), length(Seq, Length).

:- collatz(6, Seq), writeln(Seq).
:- collatz_length(27, Len), writeln(Len).
